### PR TITLE
Fix data corruption in PostgreSQL tables with geometry attributes

### PR DESCRIPTION
**Pull request by strk**
References #49380

### CODE REVIEW

**Comment by strk** [comment]
@MorriganR comments in this PR are highly welcome, thanks

**Comment by MorriganR** [comment]
Everything seems to be working as intended.

I was wondering what would happen if the wkt string length limit worked.

<details>
<summary>data set</summary>

````sql
DROP TABLE IF EXISTS public.test_geom;
CREATE TABLE IF NOT EXISTS public.test_geom
(
    pk integer NOT NULL,
    test_int integer DEFAULT 0,
    geom geometry(Point,4326),
    geom_srid geometry(Point,4326),
    geom_no_srid geometry(Point),
    geom_no_srid_line geometry(Linestring),
    CONSTRAINT test_geom_pkey PRIMARY KEY (pk)
)

DELETE FROM public.test_geom;
INSERT INTO public.test_geom (pk, test_int, geom, geom_srid, geom_no_srid, geom_no_srid_line)
  VALUES (1, 0,
    ST_GeomFromEWKT('SRID=4326;POINT(1.0 1.0)'),
    ST_GeomFromEWKT('SRID=4326;POINT(1.0 1.0)'),
    ST_GeomFromEWKT('POINT(1.0 1.0)'),
    (SELECT ST_GeomFromEWKT('LINESTRING(' || array_to_string(
      ARRAY(
        SELECT 1.0+dx/1000.0 || ' ' || 1.0+dy/1000.0 FROM generate_series(1,33) dx, generate_series(1,33) dy
      ), ', ') || ')'))
  );

SELECT pk, test_int, st_asewkt(geom) as geom,
  st_asewkt(geom_srid) as geom_srid,
  st_asewkt(geom_no_srid) as geom_no_srid,
  st_asewkt(geom_no_srid_line) as geom_no_srid_line
  FROM public.test_geom;
````

</details>

I was very surprised to get the "trimmed" geometry without any errors or warnings 8)

init state


after set `test_int` = 42 & save

**Comment by strk** [comment]
I've actually changed the code to NOT trim that value anymore, are you still getting the trims ?
Do you have powers to review/approve this PR ? It would help merging it in :)

**Comment by MorriganR** [comment]
Sorry, no credentials :)
I'm not a wizard, I'm just learning...

**Comment by MorriganR** [comment]
> I've actually changed the code to NOT trim that value anymore, are you still getting the trims ?

You are absolutely right! I didn't pay attention, the field is truncated with a length of 32767 characters, it looks like some Qt limitations...

**Comment by strk** [comment]
Ah, that explains the custom trimming then. It's to have something nicer looking than the QT default!
So we're left with a TextField initialized with a value which cannot be safely read back. Only option then is to mark it as read-only.

**Comment by strk** [comment]
I think the problem is that there's some code path that BEFORE sending the value to the data provider, "tweaks" it while still inside the form, and it could be `src/gui/qgsattributeform.cpp:446 : (saveEdits)` I saw for example how removing a single ordinate value from a LINESTRING results in dropping the whole coordinate, or how cutting the final part of the WKT re-adds the required closing parenthesis. I'm not sure how that works and what's responsible for that but it may explain why you may get unwanted edits if you touch the long WKT

**Comment by m-kuhn** [comment]
IMO we should disable editing of referenced geometries via text edit. If secondary geom columns should be editable, a dedicated widget would be more appropriate.

**Comment by strk** [comment]
Alright, commit [b1dc040](https://github.com/qgis/QGIS/pull/50592/commits/b1dc04048c6e5f0be7ead0a395e3a8b5328a39bb) disables editing of referenced geometries via text field

**Comment by strk** [comment]
Bah, I'm not convinced about the last commit. In the previous state we were after all able to edit short geometries, I'd revert back.

**Comment by strk** [comment]
I've reverted to still allow editing. It's a step forward from the corruption in that if you don't try to edit the values, you don't get the corruption (previously you'd still get the corruption by editing a different field). There are also other small improvement in this PR which seem good to me, although I'm not very expert in this part of the codebase.

**Comment by strk** [comment]
@MorriganR can you try again with current code ? I believe  you shouldn't get the truncation if you don't touch the long field but only touch the integer field.

**Comment by MorriganR** [comment]
truncation - fixed
<details>



</details>

File: src/core/qgsfield.cpp [hunk 1]
```diff
@@ -589,12 +589,61 @@ bool QgsField::convertCompatible( QVariant &v, QString *errorMessage ) const
     return false;
   }
 
+  if ( d->typeName == "geometry" && v.type() == QVariant::String )
+  {
+    // TODO: expose this as a "canonical text input" method of
```
Comment (lbartoletti) [inline src/core/qgsfield.cpp:594]: Is this part could be useful in a dedicated method?

File: src/core/qgsfield.cpp [hunk 2]
```diff
@@ -589,12 +589,61 @@ bool QgsField::convertCompatible( QVariant &v, QString *errorMessage ) const
...
+    // QgsReferencedGeometry class
+    auto parts = v.toString().split( " [" );
+    auto wkt = parts[0];
+    // TODO: handle more string to mean NULL ?
```
Comment (lbartoletti) [inline src/core/qgsfield.cpp:598]: I don't understand this first test. Why don't try to use directly `QgsGeometry::fromWkt`?

Comment (MorriganR) [reply src/core/qgsfield.cpp:598]: `Point (1 1) [EPSG:4326 - WGS 84]` - not similar to WKT

Comment (lbartoletti) [reply src/core/qgsfield.cpp:598]: Oh, ok!
It is a (yet another) normalized way to write this EWKT-like geometry ?

**Comment by m-kuhn** [comment]
IIRC: EWKT is a postgis internal format, not used by any other data provider, the SRID is specific to the database we are connected to, so the string is not portable.
Is it worth an effort to expose this to a user, in a limited way (editing such a string by hand ...), risking performance problems, ... ?

File: src/core/qgsfield.cpp (hunk 2 above)
Comment (MorriganR) [reply src/core/qgsfield.cpp:598]: this format was introduced in #45127

Comment (m-kuhn) [reply src/core/qgsfield.cpp:598]: ... which was meant to be representation only ...

**Comment by elpaso** [comment]
> ~IIRC: EWKT is a postgis internal format, not used by any other data provider, the SRID is specific to the database we are connected to, so the string is not portable.~ The format implemented in #45127 has been thought as a human readable display-only implementation. Is it worth an effort to expose this editable to a user? Editing such a string by hand is cumbersome, performance limitations may apply (huge geometry strings), complete reversibility with all sorts of CRS would need to be tested in depth, ...

IIRC there were also problems with the dialog becoming too wide with unlimited string widths and even crashing on some display managers but I might be confusing this with another issue.

**Comment by MorriganR** [comment]
"Data types supported by the PostgreSQL provider include:
integer, float, boolean, binary object, varchar, **geometry**, timestamp,
array, hstore and json."
https://github.com/qgis/QGIS-Documentation/blame/master/docs/user_manual/managing_data_source/supported_data.rst#L392-L394

File: src/core/qgsfield.cpp (hunk 1 above)
Comment (strk) [reply src/core/qgsfield.cpp:594]: Yes, I think "canonical text input and output" methods should be exposed for the ReferencedGeometry class

**Comment by strk** [comment]
I would note that for points most problems listed are not really problems. The long-string issue is currently handled by truncating the WKT which effectively results in making the field value "fragile" (changes drastically if you try to edit it).

Should we decide to make the text-widget read-only where should such choice be made, in code ?

**Comment by elpaso** [comment]
> I would note that for points most problems listed are not really problems. The long-string issue is currently handled by truncating the WKT which effectively results in making the field value "fragile" (changes drastically if you try to edit it).
> 
> Should we decide to make the text-widget read-only where should such choice be made, in code ?

Well, ideally we should use the digitizing methods available in QGIS to edit a (secondary) geometry column. It seems to me more a feature request than a bug though.

IMHO manual editing a WKT from a text line edit is something I don't think we should support.

**Comment by strk** [comment]
Yes editing with digitizing tools would ideal, and a feature. This PR is aimed at fixing a data corruption bug though and it is reported to be effective at that.

Also, QGIS-3.22 allowed to edit secondary geometries in text form (with whatever format the provider used as the canonical text format for the geometry) so preventing that would be a regression.

**Comment by strk** [comment]
I just tried QGIS-3.22 again to make sure I didn't overlook it and I do confirm that I can edit the secondary geometries in WKT format just fine, including changing the SRID. No truncation at all occurs there, so even with the code in this PR we do have the regression of having the text truncated (thus not allowing full edit).

It's not necessarely "manually" that you edit those strings, could be a copy-paste from another place...

**Comment by strk** [comment]
I've noticed that despite QgsFieldValidator::validate returning Invalid for user-entered values those are still remembered by the form, is that expected ?

**Comment by strk** [comment]
Are there examples of data type for which editing is disabled with text editor ?

**Comment by strk** [comment]
@nyalldawson suggested looking at his geometry widget in https://github.com/nyalldawson/QGIS/commit/8bf716e63eaf87e4e01742e136dd4ce086d34c7f

**Comment by strk** [comment]
@MorriganR it looks like this fix didn't make it into the release, or did 3.28 came out with the regression bug fixed ?

**Comment by strk** [comment]
This change fixes a data corruption bug. It's not perfect, but allows users to NOT corrupt their data by simply modifying an attribute

**Comment by strk** [comment]
@nyalldawson as we are in feature freeze mode again, I dubt geom editing widget would fit, while data corruption bug is probably still there ( #49380 ) -- I'll verify if the bug is still there and if this fix still fixes it, but if it does, would you be ok with merging it even if sub-optimal as a fix ? If not else we'd let people possibly edit their data w/out loosing the geometric component

**Comment by strk** [comment]
Rebased and re-tested. I confirm that bug #49380 still exists in master branch and that this pull request fixes it.
At least with geometries having a short WKT representation.

**Comment by strk** [comment]
@nyalldawson I took at look at your https://github.com/qgis/QGIS/compare/master...nyalldawson:QGIS:geo_widget2 but it didn't seem to have made it into QGIS official branch, am I right ?

**Comment by nyalldawson** [comment]
@strk 
> but it didn't seem to have made it into QGIS official branch, am I right ?

Correct -- it's an ~8 year old proof of concept, but I haven't had a use case for multi-geometry columns in years and so never bothered to get it merge ready.

**Comment by nyalldawson** [comment]
See https://github.com/qgis/QGIS/pull/51863

**Comment by strk** [comment]
Thanks @nyalldawson and @elpaso and @MorriganR .. I've rebased this work against current master, there doesn't seem to be much left except the QgsPostgresProvider::fromEwkt function fix to avoid discarding geometries with no SRIDs, should I turn this PR into something addressing that problem ? Or is that problem already fixed now, @MorriganR ?

I'll build the 3 branches (master, this, and @nyalldawson one from PR #51863 to check with my eyes, but it will surely take time)

**Comment by strk** [comment]
I've verified #51863 prevents the data corruption, although it still doesn't allow editing the geometries.

**Comment by nyalldawson** [comment]
@strk this can be closed now, right?

**Comment by strk** [comment]
Yeah, I was checking exactly this. Closing.

**Pull request closed without merging**